File: hubz-backend/src/test/java/com/hubz/application/service/OrganizationServiceTest.java
package com.hubz.application.service;

import com.hubz.application.dto.request.CreateOrganizationRequest;
import com.hubz.application.dto.request.UpdateOrganizationRequest;
import com.hubz.application.dto.response.MemberResponse;
import com.hubz.application.dto.response.OrganizationResponse;
import com.hubz.application.port.out.OrganizationMemberRepositoryPort;
import com.hubz.application.port.out.OrganizationRepositoryPort;
import com.hubz.application.port.out.UserRepositoryPort;
import com.hubz.domain.enums.MemberRole;
import com.hubz.domain.exception.MemberAlreadyExistsException;
import com.hubz.domain.exception.OrganizationNotFoundException;
import com.hubz.domain.exception.UserNotFoundException;
import com.hubz.domain.model.Organization;
import com.hubz.domain.model.OrganizationMember;
import com.hubz.domain.model.User;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
@DisplayName("OrganizationService Unit Tests")
class OrganizationServiceTest {

    @Mock
    private OrganizationRepositoryPort organizationRepository;

    @Mock
    private OrganizationMemberRepositoryPort memberRepository;

    @Mock
    private UserRepositoryPort userRepository;

    @Mock
    private AuthorizationService authorizationService;

    @InjectMocks
    private OrganizationService organizationService;

    private UUID ownerId;
    private UUID userId;
    private Organization testOrg;
    private User testUser;
    private OrganizationMember testMember;
    private CreateOrganizationRequest createRequest;
    private UpdateOrganizationRequest updateRequest;

    @BeforeEach
    void setUp() {
        ownerId = UUID.randomUUID();
        userId = UUID.randomUUID();

        testOrg = Organization.builder()
                .id(UUID.randomUUID())
                .name("Test Organization")
                .description("Test description")
                .icon("icon.png")
                .color("#FF5733")
                .ownerId(ownerId)
                .createdAt(LocalDateTime.now())
                .build();

        testUser = User.builder()
                .id(userId)
                .email("test@example.com")
                .password("encodedPassword")
                .firstName("John")
                .lastName("Doe")
                .description("Test user")
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        testMember = OrganizationMember.builder()
                .id(UUID.randomUUID())
                .organizationId(testOrg.getId())
                .userId(userId)
                .role(MemberRole.MEMBER)
                .joinedAt(LocalDateTime.now())
                .build();

        createRequest = new CreateOrganizationRequest(
                "Test Organization",
                "Test description",
                "icon.png",
                "#FF5733"
        );

        updateRequest = new UpdateOrganizationRequest(
                "Updated Organization",
                "Updated description",
                "newicon.png",
                "#00FF00",
                "# README"
        );
    }

    @Nested
    @DisplayName("Create Organization Tests")
    class CreateTests {

        @Test
        @DisplayName("Should successfully create organization")
        void shouldCreateOrganization() {
            // Given
            when(organizationRepository.save(any(Organization.class))).thenReturn(testOrg);
            when(memberRepository.save(any(OrganizationMember.class))).thenReturn(testMember);

            // When
            OrganizationResponse response = organizationService.create(createRequest, ownerId);

            // Then
            assertThat(response).isNotNull();
            assertThat(response.getName()).isEqualTo(testOrg.getName());
            assertThat(response.getDescription()).isEqualTo(testOrg.getDescription());
            assertThat(response.getIcon()).isEqualTo(testOrg.getIcon());
            assertThat(response.getColor()).isEqualTo(testOrg.getColor());
            assertThat(response.getOwnerId()).isEqualTo(ownerId);

            verify(organizationRepository).save(any(Organization.class));
            verify(memberRepository).save(any(OrganizationMember.class));
        }

        @Test
        @DisplayName("Should create owner member when creating organization")
        void shouldCreateOwnerMember() {
            // Given
            when(organizationRepository.save(any(Organization.class))).thenReturn(testOrg);
            ArgumentCaptor<OrganizationMember> memberCaptor = ArgumentCaptor.forClass(OrganizationMember.class);

            // When
            organizationService.create(createRequest, ownerId);

            // Then
            verify(memberRepository).save(memberCaptor.capture());
            OrganizationMember savedMember = memberCaptor.getValue();
            assertThat(savedMember.getOrganizationId()).isEqualTo(testOrg.getId());
            assertThat(savedMember.getUserId()).isEqualTo(ownerId);
            assertThat(savedMember.getRole()).isEqualTo(MemberRole.OWNER);
            assertThat(savedMember.getJoinedAt()).isNotNull();
        }

        @Test
        @DisplayName("Should set createdAt timestamp")
        void shouldSetCreatedAt() {
            // Given
            when(organizationRepository.save(any(Organization.class))).thenReturn(testOrg);
            ArgumentCaptor<Organization> orgCaptor = ArgumentCaptor.forClass(Organization.class);

            // When
            organizationService.create(createRequest, ownerId);

            // Then
            verify(organizationRepository).save(orgCaptor.capture());
            Organization savedOrg = orgCaptor.getValue();
            assertThat(savedOrg.getCreatedAt()).isNotNull();
        }

        @Test
        @DisplayName("Should generate UUIDs for organization and member")
        void shouldGenerateUUIDs() {
            // Given
            when(organizationRepository.save(any(Organization.class))).thenReturn(testOrg);
            ArgumentCaptor<Organization> orgCaptor = ArgumentCaptor.forClass(Organization.class);
            ArgumentCaptor<OrganizationMember> memberCaptor = ArgumentCaptor.forClass(OrganizationMember.class);

            // When
            organizationService.create(createRequest, ownerId);

            // Then
            verify(organizationRepository).save(orgCaptor.capture());
            verify(memberRepository).save(memberCaptor.capture());
            assertThat(orgCaptor.getValue().getId()).isNotNull();
            assertThat(memberCaptor.getValue().getId()).isNotNull();
        }
    }

    @Nested
    @DisplayName("Get All Organizations Tests")
    class GetAllTests {

        @Test
        @DisplayName("Should return list of all organizations")
        void shouldGetAllOrganizations() {
            // Given
            List<Organization> orgs = List.of(testOrg);
            when(organizationRepository.findAll()).thenReturn(orgs);

            // When
            List<OrganizationResponse> responses = organizationService.getAll();

            // Then
            assertThat(responses).hasSize(1);
            assertThat(responses.get(0).getName()).isEqualTo(testOrg.getName());
            verify(organizationRepository).findAll();
        }

        @Test
        @DisplayName("Should return empty list when no organizations exist")
        void shouldReturnEmptyList() {
            // Given
            when(organizationRepository.findAll()).thenReturn(List.of());

            // When
            List<OrganizationResponse> responses = organizationService.getAll();

            // Then
            assertThat(responses).isEmpty();
            verify(organizationRepository).findAll();
        }
    }

    @Nested
    @DisplayName("Get Organization By ID Tests")
    class GetByIdTests {

        @Test
        @DisplayName("Should successfully get organization by ID")
        void shouldGetOrganizationById() {
            // Given
            when(organizationRepository.findById(testOrg.getId())).thenReturn(Optional.of(testOrg));

            // When
            OrganizationResponse response = organizationService.getById(testOrg.getId());

            // Then
            assertThat(response).isNotNull();
            assertThat(response.getId()).isEqualTo(testOrg.getId());
            assertThat(response.getName()).isEqualTo(testOrg.getName());
            verify(organizationRepository).findById(testOrg.getId());
        }

        @Test
        @DisplayName("Should throw OrganizationNotFoundException when not found")
        void shouldThrowExceptionWhenNotFound() {
            // Given
            UUID nonExistentId = UUID.randomUUID();
            when(organizationRepository.findById(nonExistentId)).thenReturn(Optional.empty());

            // When & Then
            assertThatThrownBy(() -> organizationService.getById(nonExistentId))
                    .isInstanceOf(OrganizationNotFoundException.class);
            verify(organizationRepository).findById(nonExistentId);
        }
    }

    @Nested
    @DisplayName("Update Organization Tests")
    class UpdateTests {

        @Test
        @DisplayName("Should successfully update organization")
        void shouldUpdateOrganization() {
            // Given
            doNothing().when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            when(organizationRepository.findById(testOrg.getId())).thenReturn(Optional.of(testOrg));
            when(organizationRepository.save(any(Organization.class))).thenReturn(testOrg);

            // When
            OrganizationResponse response = organizationService.update(testOrg.getId(), updateRequest, ownerId);

            // Then
            assertThat(response).isNotNull();
            verify(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            verify(organizationRepository).findById(testOrg.getId());
            verify(organizationRepository).save(any(Organization.class));
        }

        @Test
        @DisplayName("Should throw exception when user is not admin")
        void shouldThrowExceptionWhenNotAdmin() {
            // Given
            UUID nonAdminUserId = UUID.randomUUID();
            doThrow(new RuntimeException("Not admin"))
                    .when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), nonAdminUserId);

            // When & Then
            assertThatThrownBy(() -> organizationService.update(testOrg.getId(), updateRequest, nonAdminUserId))
                    .isInstanceOf(RuntimeException.class);
            verify(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), nonAdminUserId);
            verify(organizationRepository, never()).save(any());
        }

        @Test
        @DisplayName("Should throw exception when organization not found")
        void shouldThrowExceptionWhenOrganizationNotFound() {
            // Given
            UUID nonExistentId = UUID.randomUUID();
            doNothing().when(authorizationService).checkOrganizationAdminAccess(nonExistentId, ownerId);
            when(organizationRepository.findById(nonExistentId)).thenReturn(Optional.empty());

            // When & Then
            assertThatThrownBy(() -> organizationService.update(nonExistentId, updateRequest, ownerId))
                    .isInstanceOf(OrganizationNotFoundException.class);
            verify(organizationRepository, never()).save(any());
        }
    }

    @Nested
    @DisplayName("Delete Organization Tests")
    class DeleteTests {

        @Test
        @DisplayName("Should successfully delete organization")
        void shouldDeleteOrganization() {
            // Given
            doNothing().when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            when(organizationRepository.findById(testOrg.getId())).thenReturn(Optional.of(testOrg));
            doNothing().when(organizationRepository).deleteById(testOrg.getId());

            // When
            organizationService.delete(testOrg.getId(), ownerId);

            // Then
            verify(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            verify(organizationRepository).findById(testOrg.getId());
            verify(organizationRepository).deleteById(testOrg.getId());
        }

        @Test
        @DisplayName("Should throw exception when user is not admin")
        void shouldThrowExceptionWhenNotAdmin() {
            // Given
            UUID nonAdminUserId = UUID.randomUUID();
            doThrow(new RuntimeException("Not admin"))
                    .when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), nonAdminUserId);

            // When & Then
            assertThatThrownBy(() -> organizationService.delete(testOrg.getId(), nonAdminUserId))
                    .isInstanceOf(RuntimeException.class);
            verify(organizationRepository, never()).deleteById(any());
        }

        @Test
        @DisplayName("Should throw exception when organization not found")
        void shouldThrowExceptionWhenOrganizationNotFound() {
            // Given
            UUID nonExistentId = UUID.randomUUID();
            doNothing().when(authorizationService).checkOrganizationAdminAccess(nonExistentId, ownerId);
            when(organizationRepository.findById(nonExistentId)).thenReturn(Optional.empty());

            // When & Then
            assertThatThrownBy(() -> organizationService.delete(nonExistentId, ownerId))
                    .isInstanceOf(OrganizationNotFoundException.class);
            verify(organizationRepository, never()).deleteById(any());
        }
    }

    @Nested
    @DisplayName("Get Members Tests")
    class GetMembersTests {

        @Test
        @DisplayName("Should successfully get organization members")
        void shouldGetMembers() {
            // Given
            doNothing().when(authorizationService).checkOrganizationAccess(testOrg.getId(), ownerId);
            when(organizationRepository.findById(testOrg.getId())).thenReturn(Optional.of(testOrg));
            when(memberRepository.findByOrganizationId(testOrg.getId())).thenReturn(List.of(testMember));
            when(userRepository.findById(testMember.getUserId())).thenReturn(Optional.of(testUser));

            // When
            List<MemberResponse> members = organizationService.getMembers(testOrg.getId(), ownerId);

            // Then
            assertThat(members).hasSize(1);
            assertThat(members.get(0).getUserId()).isEqualTo(testUser.getId());
            assertThat(members.get(0).getEmail()).isEqualTo(testUser.getEmail());
            verify(authorizationService).checkOrganizationAccess(testOrg.getId(), ownerId);
            verify(memberRepository).findByOrganizationId(testOrg.getId());
        }

        @Test
        @DisplayName("Should throw exception when organization not found")
        void shouldThrowExceptionWhenOrganizationNotFound() {
            // Given
            UUID nonExistentId = UUID.randomUUID();
            doNothing().when(authorizationService).checkOrganizationAccess(nonExistentId, ownerId);
            when(organizationRepository.findById(nonExistentId)).thenReturn(Optional.empty());

            // When & Then
            assertThatThrownBy(() -> organizationService.getMembers(nonExistentId, ownerId))
                    .isInstanceOf(OrganizationNotFoundException.class);
            verify(memberRepository, never()).findByOrganizationId(any());
        }
    }

    @Nested
    @DisplayName("Add Member Tests")
    class AddMemberTests {

        @Test
        @DisplayName("Should successfully add member to organization")
        void shouldAddMember() {
            // Given
            UUID newUserId = UUID.randomUUID();
            doNothing().when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            when(organizationRepository.findById(testOrg.getId())).thenReturn(Optional.of(testOrg));
            when(memberRepository.existsByOrganizationIdAndUserId(testOrg.getId(), newUserId)).thenReturn(false);
            when(userRepository.findById(newUserId)).thenReturn(Optional.of(testUser));
            when(memberRepository.save(any(OrganizationMember.class))).thenReturn(testMember);

            // When
            MemberResponse response = organizationService.addMember(testOrg.getId(), newUserId, MemberRole.MEMBER, ownerId);

            // Then
            assertThat(response).isNotNull();
            assertThat(response.getUserId()).isEqualTo(testUser.getId());
            assertThat(response.getRole()).isEqualTo(MemberRole.MEMBER);
            verify(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            verify(memberRepository).save(any(OrganizationMember.class));
        }

        @Test
        @DisplayName("Should throw exception when member already exists")
        void shouldThrowExceptionWhenMemberAlreadyExists() {
            // Given
            doNothing().when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            when(organizationRepository.findById(testOrg.getId())).thenReturn(Optional.of(testOrg));
            when(memberRepository.existsByOrganizationIdAndUserId(testOrg.getId(), userId)).thenReturn(true);

            // When & Then
            assertThatThrownBy(() -> organizationService.addMember(testOrg.getId(), userId, MemberRole.MEMBER, ownerId))
                    .isInstanceOf(MemberAlreadyExistsException.class);
            verify(memberRepository, never()).save(any());
        }

        @Test
        @DisplayName("Should throw exception when user not found")
        void shouldThrowExceptionWhenUserNotFound() {
            // Given
            UUID nonExistentUserId = UUID.randomUUID();
            doNothing().when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            when(organizationRepository.findById(testOrg.getId())).thenReturn(Optional.of(testOrg));
            when(memberRepository.existsByOrganizationIdAndUserId(testOrg.getId(), nonExistentUserId)).thenReturn(false);
            when(userRepository.findById(nonExistentUserId)).thenReturn(Optional.empty());

            // When & Then
            assertThatThrownBy(() -> organizationService.addMember(testOrg.getId(), nonExistentUserId, MemberRole.MEMBER, ownerId))
                    .isInstanceOf(UserNotFoundException.class);
            verify(memberRepository, never()).save(any());
        }
    }

    @Nested
    @DisplayName("Remove Member Tests")
    class RemoveMemberTests {

        @Test
        @DisplayName("Should successfully remove member from organization")
        void shouldRemoveMember() {
            // Given
            doNothing().when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            doNothing().when(memberRepository).deleteByOrganizationIdAndUserId(testOrg.getId(), userId);

            // When
            organizationService.removeMember(testOrg.getId(), userId, ownerId);

            // Then
            verify(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), ownerId);
            verify(memberRepository).deleteByOrganizationIdAndUserId(testOrg.getId(), userId);
        }

        @Test
        @DisplayName("Should throw exception when user is not admin")
        void shouldThrowExceptionWhenNotAdmin() {
            // Given
            UUID nonAdminUserId = UUID.randomUUID();
            doThrow(new RuntimeException("Not admin"))
                    .when(authorizationService).checkOrganizationAdminAccess(testOrg.getId(), nonAdminUserId);

            // When & Then
            assertThatThrownBy(() -> organizationService.removeMember(testOrg.getId(), userId, nonAdminUserId))
                    .isInstanceOf(RuntimeException.class);
            verify(memberRepository, never()).deleteByOrganizationIdAndUserId(any(), any());
        }
    }
}

File: hubz-frontend/src/services/organization.service.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { organizationService } from './organization.service';
import type { Organization, Member } from '../types/organization';

// Mock the api module
vi.mock('./api', () => ({
  default: {
    get: vi.fn(),
    post: vi.fn(),
    put: vi.fn(),
    delete: vi.fn(),
    patch: vi.fn(),
  },
}));

import api from './api';

describe('organizationService', () => {
  const mockOrganization: Organization = {
    id: '123e4567-e89b-12d3-a456-426614174000',
    name: 'Test Org',
    description: 'A test organization',
    icon: 'building',
    color: '#3B82F6',
    readme: null,
    logoUrl: null,
    ownerId: '456e4567-e89b-12d3-a456-426614174000',
    createdAt: '2024-01-01T00:00:00Z',
  };

  const mockMember: Member = {
    id: 'member-1',
    userId: 'user-1',
    firstName: 'Jane',
    lastName: 'Doe',
    email: 'jane@example.com',
    profilePhotoUrl: null,
    role: 'MEMBER',
    joinedAt: '2024-01-01T00:00:00Z',
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('getAll', () => {
    it('should fetch all organizations', async () => {
      const orgs = [mockOrganization];
      vi.mocked(api.get).mockResolvedValueOnce({ data: orgs });

      const result = await organizationService.getAll();

      expect(api.get).toHaveBeenCalledWith('/organizations');
      expect(result).toEqual(orgs);
    });

    it('should return empty array when no organizations', async () => {
      vi.mocked(api.get).mockResolvedValueOnce({ data: [] });

      const result = await organizationService.getAll();

      expect(result).toEqual([]);
    });
  });

  describe('getById', () => {
    it('should fetch organization by id', async () => {
      vi.mocked(api.get).mockResolvedValueOnce({ data: mockOrganization });

      const result = await organizationService.getById('123');

      expect(api.get).toHaveBeenCalledWith('/organizations/123');
      expect(result).toEqual(mockOrganization);
    });
  });

  describe('create', () => {
    it('should create a new organization', async () => {
      vi.mocked(api.post).mockResolvedValueOnce({ data: mockOrganization });

      const result = await organizationService.create({
        name: 'Test Org',
        description: 'A test organization',
      });

      expect(api.post).toHaveBeenCalledWith('/organizations', {
        name: 'Test Org',
        description: 'A test organization',
      });
      expect(result).toEqual(mockOrganization);
    });
  });

  describe('update', () => {
    it('should update an organization', async () => {
      const updated = { ...mockOrganization, name: 'Updated Org' };
      vi.mocked(api.put).mockResolvedValueOnce({ data: updated });

      const result = await organizationService.update('123', { name: 'Updated Org' });

      expect(api.put).toHaveBeenCalledWith('/organizations/123', { name: 'Updated Org' });
      expect(result.name).toBe('Updated Org');
    });
  });

  describe('delete', () => {
    it('should delete an organization', async () => {
      vi.mocked(api.delete).mockResolvedValueOnce({ data: undefined });

      await organizationService.delete('123');

      expect(api.delete).toHaveBeenCalledWith('/organizations/123');
    });
  });

  describe('getMembers', () => {
    it('should fetch members of an organization', async () => {
      const members = [mockMember];
      vi.mocked(api.get).mockResolvedValueOnce({ data: members });

      const result = await organizationService.getMembers('org-123');

      expect(api.get).toHaveBeenCalledWith('/organizations/org-123/members');
      expect(result).toEqual(members);
    });
  });

  describe('changeMemberRole', () => {
    it('should change a member role', async () => {
      const updatedMember = { ...mockMember, role: 'ADMIN' as const };
      vi.mocked(api.patch).mockResolvedValueOnce({ data: updatedMember });

      const result = await organizationService.changeMemberRole('org-123', 'user-1', 'ADMIN');

      expect(api.patch).toHaveBeenCalledWith(
        '/organizations/org-123/members/user-1/role',
        { role: 'ADMIN' }
      );
      expect(result.role).toBe('ADMIN');
    });
  });

  describe('removeMember', () => {
    it('should remove a member from organization', async () => {
      vi.mocked(api.delete).mockResolvedValueOnce({ data: undefined });

      await organizationService.removeMember('org-123', 'user-1');

      expect(api.delete).toHaveBeenCalledWith('/organizations/org-123/members/user-1');
    });
  });

  describe('transferOwnership', () => {
    it('should transfer ownership to another user', async () => {
      vi.mocked(api.post).mockResolvedValueOnce({ data: undefined });

      await organizationService.transferOwnership('org-123', 'new-owner-id');

      expect(api.post).toHaveBeenCalledWith(
        '/organizations/org-123/transfer-ownership/new-owner-id'
      );
    });
  });

  describe('uploadLogo', () => {
    it('should upload organization logo', async () => {
      const updatedOrg = { ...mockOrganization, logoUrl: 'logos/logo.png' };
      vi.mocked(api.post).mockResolvedValueOnce({ data: updatedOrg });

      const file = new File(['logo'], 'logo.png', { type: 'image/png' });
      const result = await organizationService.uploadLogo('org-123', file);

      expect(api.post).toHaveBeenCalledWith(
        '/organizations/org-123/logo',
        expect.any(FormData),
        { headers: { 'Content-Type': 'multipart/form-data' } }
      );
      expect(result.logoUrl).toBe('logos/logo.png');
    });
  });

  describe('deleteLogo', () => {
    it('should delete organization logo', async () => {
      const updatedOrg = { ...mockOrganization, logoUrl: null };
      vi.mocked(api.delete).mockResolvedValueOnce({ data: updatedOrg });

      const result = await organizationService.deleteLogo('org-123');

      expect(api.delete).toHaveBeenCalledWith('/organizations/org-123/logo');
      expect(result.logoUrl).toBeNull();
    });
  });
});
